sympy/physics/vector/frame.py

```diff
@@ -535,20 +535,21 @@ def _dcm(self, parent, parent_orient):
         frames = self._dcm_cache.keys()
         dcm_dict_del = []
         dcm_cache_del = []
-        for frame in frames:
-            if frame in self._dcm_dict:
-                dcm_dict_del += [frame]
-            dcm_cache_del += [frame]
-        for frame in dcm_dict_del:
-            del frame._dcm_dict[self]
-        for frame in dcm_cache_del:
-            del frame._dcm_cache[self]
+        if parent in frames:
+            for frame in frames:
+                if frame in self._dcm_dict:
+                    dcm_dict_del += [frame]
+                dcm_cache_del += [frame]
+            for frame in dcm_dict_del:
+                del frame._dcm_dict[self]
+            for frame in dcm_cache_del:
+                del frame._dcm_cache[self]
```

Comment:
I personally still don't know why there is a dcm_cache and a dcm_dict. If you understand that, could you make some comments here explaining it.

Reply:
Yes, I would try

Reply:
@moorepants I think `_dcm_cache` and `_dcm_dict` both already have some comments explaining them. https://github.com/sympy/sympy/blob/484f1cd3fa1497cd9b43eadea32ce65dfb02e892/sympy/physics/vector/frame.py#L193-L195

Reply:
What does that mean and why are their two things?

Reply:
For example, this sentence "The _dcm_cache dictionary will work as the dcm cache. " means nothing. It's like saying "The cow is the cow".

Reply:
But I think I'm guessing more what may be going on.

`_dcm_dict` should only hold rotation matrices that relate `self` to immediately adjacent reference frames.

`_dcm_cache` can have everything that is in `_dcm_dict` but can also store relationships between `self` and any other reference frame that has a relative relationship (even if through a chain of reference frames).

If that is correct, then how do they interact. If you change something in `_dcm_dict` then I guess that means you have to clear anything in the cache that could be affected, or maybe it just gets wiped completely if `_dcm_dict` changes. The explanation I'm wondering about would answer these questions.

Reply:
I wonder if we have unit tests that ensure the `_dcm_dict` and `_dcm_cache` functionality works as expected.

Reply:
Here's the original commit that introduced this cache (I think): https://github.com/sympy/sympy/commit/758b7c5ec2dc3ccfb76eadb63bcf1f9d8c3d1865

Reply:
> `_dcm_dict` should only hold rotation matrices that relate `self` to immediately adjacent reference frames.

Yes that's how `_dcm_dict` is working.

> `_dcm_cache` can have everything that is in `_dcm_dict` but can also store relationships between `self` and any other reference frame that has a relative relationship (even if through a chain of reference frames).

I don't think that's the correct behavior of `_dcm_cache`.  I think `_dcm_cache` stores what's in `_dcm_dict` if that frame is not passed as parent frame but also stores `self's matrix` on which orientation is applied only if that frame is passed as parent frame.
 
```
>>> B.orient(A, 'Axis', (a, A.x))      # Statement 1
>>> C.orient(B, 'Axis', (b, B.x))      # Statement 2
>>> print(A._dcm_cache == A._dcm_dict)  # A is passed as parent in Statement 1
False
>>> print(B._dcm_cache == B._dcm_dict)   # B is passed as parent in Statement 2
False
>>> print(C._dcm_cache == C._dcm_dict)   # C is not passed as parent 
True

>>> print(A._dcm_dict)
{B: Matrix([
[1,      0,       0],
[0, cos(a), -sin(a)],
[0, sin(a),  cos(a)]])}

>>> print(A._dcm_cache) # Similar behavior is shown by B because it's not passed as parent frame.
{A: Matrix([
[1, 0, 0],
[0, 1, 0],
[0, 0, 1]]), B: Matrix([
[1,      0,       0],
[0, cos(a), -sin(a)],
[0, sin(a),  cos(a)]])}

>>> print(C._dcm_dict)
{B: Matrix([
[1,       0,      0],
[0,  cos(b), sin(b)],
[0, -sin(b), cos(b)]])}

>>> print(C._dcm_cache) # For frame C both are same since it is not passed as parent frame
{B: Matrix([
[1,       0,      0],
[0,  cos(b), sin(b)],
[0, -sin(b), cos(b)]])}
```

Reply:
> If you change something in `_dcm_dict` then I guess that means you have to clear anything in the cache that could be affected, or maybe it just gets wiped completely if `_dcm_dict` changes. The explanation I'm wondering about would answer these questions.

In `._dcm()` if we don't have relation already defined with parent
```
# Example
B.orient(A)
C.orient(A)
D.orient(C)
E.orient(C)
```
then in both `_dcm_dict` and `_dcm_cache` new relation is added.

But if a user do something like
```
B.orient(A,..)
A.orient(B,..)
```
then both `_dcm_dict` and `_dcm_cache` of A is cleared and new relation with B is added on the other hand `_dcm_dict` and `_dcm_cache` of frame B is updated with new relation with A while other relations of frame B stay intact.

Reply:
I still don't know what more than one data structure is needed.

Let's say we have a dictionary that stores the transformations of any adjacent (directly connected) frames. It works like this:

```
>>> B.orient(A)
>>> B.adjacent_relations
{A: A_dcm_B}
>>> A.adjacent_relations
{B: B_dcm_A}
>>> C.adjacent_relations
{}
>>> C.orient(A)
>>> C.adjacent_relations
{A: A_dcm_C}
>>> A.adjacent_relations
{B: B_dcm_A, C: C_dcm_A}
```

Now try to change existing relationships:

```
>>> C.orient(B)
Error! you can't do this because it closes the loop and it's ambiguous what relation to delete to open the loop
>>> A.orient(B)  # replaces the A<->B relationship, no problem
>>> A.adjacent_relations
{B: B_dcm_A_new, C: C_dcmA}
>>> B.adjacent_relations
{A: A_dcm_B_new}
>>> C.adjacent_relations
{A: A_dcm_C}
```

The `A.orient(B)` simply needs to update the `adjacent_relations` dict with the new relationship, nothing else should change.

So why is a `_dcm_cache` needed, what is that doing in this scenario?

Reply:
You wrote:

> print(A._dcm_cache) # Calculated dcm wrt C is added to _dcm_cache, no effect on _dcm_dict. 

Earlier I wrote:

> _dcm_cache can have everything that is in _dcm_dict but can also store relationships between self and any other reference frame that has a relative relationship (even if through a chain of reference frames).

So, is that not the same? That the cache stores relationships that are not adjacent? If you then change an adjacent relationship, you better wipe the cache because it could affect that.

Reply:
> I still don't know why more than one data structure is needed.

@moorepants I too believe we can merge `_dcm_cache` and `_dcm_dict`. `_dcm_cache` is only being used in `frame.dcm()`. `_dcm_cache` functionality could be added to _dcm_dict.

> The `A.orient(B)` simply needs to update the `adjacent_relations` dict with the new relationship, nothing else should change.
> So why is a `_dcm_cache` needed, what is that doing in this scenario?

`_dcm_cache`'s sole purpose is in `frame.dcm()` for which `_dcm_cache` needs to be treated as `_dcm_dict` but with extra feature of storing calculated dcm.

Reply:
> So, is that not the same? That the cache stores relationships that are not adjacent? 

`_dcm_cache` stores adjacent relationship along with those which are not adjacent but only when user explicitly calls `dcm()`. The dcm calculated between 2 frames adjacent or not is stored in `_dcm_cache` along with all adjacent ones.


> If you then change an adjacent relationship, you better wipe the cache because it could affect that.

Yes that's True and this the default behavior in this PR.